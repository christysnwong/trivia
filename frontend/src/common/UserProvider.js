import React, {useEffect, useState} from "react";
import UserContext from "./UserContext";

const demoUser = {
  id: 1,
  username: "testuser",
  first_name: "Test",
  last_name: "User",
  email: "test@test.com",
  stats: {
    level: 14,
    title: "Pro",
    "quizzes completed": 34,
    points: 3440,
    remainingPts: 60,
    levelPts: 300,
  },
};

const UserProvider = ({
  children,
  currUser = demoUser
}) => {
  
  const [currentUser, setCurrUser] = useState();

  return (
    // <UserContext.Provider value={{ currUser }}>
    <UserContext.Provider value={{ currUser, setCurrUser }}>
      {children}
    </UserContext.Provider>
  );
};

export default UserProvider;
