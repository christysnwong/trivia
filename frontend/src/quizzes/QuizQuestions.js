import { useEffect, useState, useContext, useRef } from "react";
import UserContext from "../common/UserContext";
import TriviaApi from "../api/api";

import Noty from 'noty';

const QuizQuestions = ({ num, question, next, reset, updateStats }) => {
  const [isAnswered, setIsAnswered] = useState(false);
  const [ansStyle, setAnsStyle] = useState({});
  const [secs, setSecs] = useState(0);
  const timerId = useRef();

  const { currUser } = useContext(UserContext);

  // stop timer
  const stopTimer = () => {
    clearInterval(timerId.current);
  };

  // calculate time bonus, increment score and combo if answering correctly
  // update stats total in parent component
  const calStats = (correct) => {
    let timeBonus = 0;
    let combo = 0;
    let score = 0;

    if (correct) {
      timeBonus = Math.min(Math.floor(Math.max((16 - secs) / 8, 0) * 6), 6);
      score++;
      combo++;
    }

    updateStats(secs, timeBonus, score, combo);
  };

  // stop timer and displays results by setting button styles - red / green
  // call calStats if answering corectly
  const checkAns = (evt) => {
    if (!isAnswered) {
      stopTimer();
      setIsAnswered(true);

      let correct = false;

      if (evt.target.id === question.correct_answer) {
        correct = true;
        setAnsStyle((old) => {
          return {
            ...old,
            [evt.target.id]: "btn-success",
          };
        });
      } else {
        setAnsStyle((old) => {
          return {
            ...old,
            [evt.target.id]: "btn-danger",
            [question.correct_answer]: "btn-success",
          };
        });
      }

      calStats(correct);
    }
  };

  // set answering status, resets timer and moves on to next question
  const nextQ = () => {
    setIsAnswered(false);
    setSecs(0);

    next();
  };

  // add trivia question to user's favourites
  const addToFav = async () => {

    try {
      let resp = await TriviaApi.addToFav(currUser.username, {
        userId: currUser.id,
        question: question.question,
        answer: question.correct_answer,
        folderName: "All",
      });

      if (resp.added) {

        new Noty({
          type: "success",
          layout: "topRight",
          theme: "bootstrap-v4",
          text: `You have successfully added this to your favourite. 
              You can revisit this in folder 'All' under 'My Favourites'.`,
          timeout: "3000",
          progressBar: true,
          closeWith: ["click"],
          killer: true,
        }).show();
      } 

    } catch (e) {
      console.error(e);

      new Noty({
        type: "warning",
        layout: "topRight",
        theme: "bootstrap-v4",
        text: `This has already been added to your favourite'.`,
        timeout: "3000",
        progressBar: true,
        closeWith: ["click"],
        killer: true,
      }).show();
    }
     
  };

  // create an object to map button styles 
  // reload when answering status and quiz data change
  useEffect(() => {

    const genMap = (answers) => {
      const result = {};
      answers.forEach((ans) => {
        result[ans] = "btn-outline-secondary";
      });

      return result;
    };

    const classMap = genMap(question.answers);

    if (!isAnswered) setAnsStyle(classMap);

  }, [isAnswered, question.answers]);

  // clean up and runs timer when quiz data changes
  useEffect(() => {
    timerId.current = setInterval(() => {
      setSecs((secs) => secs + 1);
    }, 1000);

    return () => {
      clearInterval(timerId.current);
    };
  }, [question.answers]);

  return (
    <>
      
      <p>Time Duration: {secs} secs</p>
      <hr />
      <h5 className="mb-4">
        Question {num + 1}:{" "}
        <span dangerouslySetInnerHTML={{ __html: question.question }}></span>
      </h5>

      {question.answers.map((ans, idx) => (
        <p className="btn-ans" key={idx}>
          <button
            id={ans}
            className={`btn ${ansStyle[ans]}`}
            onClick={checkAns}
            dangerouslySetInnerHTML={{ __html: ans }}
          ></button>
        </p>
      ))}

      <hr />
      <div>
        <button className="btn btn-secondary me-3" onClick={reset}>
          Reset
        </button>

        {isAnswered && currUser && (
          <button className="btn btn-warning me-3" onClick={addToFav}>
            Add to Favourite
          </button>
        )}

        {isAnswered && (
          <button className="btn btn-primary me-3" onClick={nextQ}>
            Next
          </button>
        )}
      </div>
    </>
  );
};

export default QuizQuestions;
