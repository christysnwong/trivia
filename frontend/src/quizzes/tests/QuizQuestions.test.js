import React from "react";
import { render } from "@testing-library/react";
import QuizQuestions from "../QuizQuestions";
import { MemoryRouter } from "react-router-dom";
import UserProvider from "../../common/UserProvider";

const question = {
  category: "General Knowledge",
  type: "multiple",
  difficulty: "easy",
  question:
    "Virgin Trains, Virgin Atlantic and Virgin Racing, are all companies owned by which famous entrepreneur?   ",
  correct_answer: "Richard Branson",
  incorrect_answers: ["Alan Sugar", "Donald Trump", "Bill Gates"],
  answers: ["Richard Branson", "Bill Gates", "Alan Sugar", "Donald Trump"],
};


it("renders without crashing", function () {
  render(
    <UserProvider>
      <QuizQuestions num={1} question={question} />
    </UserProvider>
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <QuizQuestions num={1} question={question}/>
      </UserProvider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});
