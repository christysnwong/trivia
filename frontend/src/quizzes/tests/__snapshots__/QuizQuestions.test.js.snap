// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <p>
    Time Duration: 0 secs
  </p>
  <hr />
  <h5
    class="mb-4"
  >
    Question 2: 
    <span>
      Virgin Trains, Virgin Atlantic and Virgin Racing, are all companies owned by which famous entrepreneur?   
    </span>
  </h5>
  <p
    class="btn-ans"
  >
    <button
      class="btn btn-outline-secondary"
      id="Richard Branson"
    >
      Richard Branson
    </button>
  </p>
  <p
    class="btn-ans"
  >
    <button
      class="btn btn-outline-secondary"
      id="Bill Gates"
    >
      Bill Gates
    </button>
  </p>
  <p
    class="btn-ans"
  >
    <button
      class="btn btn-outline-secondary"
      id="Alan Sugar"
    >
      Alan Sugar
    </button>
  </p>
  <p
    class="btn-ans"
  >
    <button
      class="btn btn-outline-secondary"
      id="Donald Trump"
    >
      Donald Trump
    </button>
  </p>
  <hr />
  <div>
    <button
      class="btn btn-secondary me-3"
    >
      Reset
    </button>
  </div>
</DocumentFragment>
`;
